pkg/reporter/github_reporter.go

```diff
@@ -0,0 +1,322 @@
+	UnknownStatus:   "⛔️ UNKNOWN",
+}
+
+var (
+	tildeChanged = regexp.MustCompile(
+		"(?m)" + // enable multi-line mode
+			"^([\t ]*)" + // only match tilde at start of line, can lead with tabs or spaces
+			"([~])") // tilde represents changes and needs switched to exclamation for git diff
+
+	swapLeadingWhitespace = regexp.MustCompile(
+		"(?m)" + // enable multi-line mode
+			"^([\t ]*)" + // only match tilde at start of line, can lead with tabs or spaces
+			`((\-(\/\+)*)|(\+(\/\-)*)|(!))`) // match characters to swap whitespace for git diff (+, +/-, -, -/+, !)
+)
+
+// GitHubReporter implements the reporter interface for writing GitHub comments.
+type GitHubReporter struct {
+	gitHubClient github.GitHub
+	flags        flags.GitHubFlags
+	commentID    int64
+	logURL       string
+}
+
+// NewGitHubReporter creates a new GitHubReporter.
+func NewGitHubReporter(ctx context.Context, f flags.GitHubFlags) (Reporter, error) {
+	logger := logging.FromContext(ctx)
+
+	ghCtx, err := githubactions.New().Context()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create github context: %w", err)
+	}
+
+	f.FromGitHubContext(ghCtx)
+
+	if err := validateFlags(f); err != nil {
+		return nil, fmt.Errorf("failed to validate github flags: %w", err)
+	}
+
+	tokenSource, err := f.TokenSource(ctx, map[string]string{
+		"contents":      "read",
+		"pull_requests": "write",
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get token source: %w", err)
+	}
+	token, err := tokenSource.GitHubToken(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get token: %w", err)
+	}
+
+	gc := github.NewClient(
+		ctx,
+		token,
+		github.WithRetryInitialDelay(1*time.Second),
+		github.WithRetryMaxAttempts(3),
+		github.WithRetryMaxDelay(30*time.Second),
+	)
+
+	if f.FlagGitHubPullRequestNumber <= 0 {
+		prResponse, err := gc.ListPullRequestsForCommit(ctx, f.FlagGitHubOwner, f.FlagGitHubRepo, f.FlagGitHubSHA, nil)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get pull request number for commit sha: %w", err)
+		}
+
+		if len(prResponse.PullRequests) == 0 {
+			return nil, fmt.Errorf("no pull requests found for commit sha: %s", f.FlagGitHubSHA)
+		}
+
+		f.FlagGitHubPullRequestNumber = prResponse.PullRequests[0].Number
+	}
+
+	logger.DebugContext(ctx, "computed pull request number", "computed_pull_request_number", f.FlagGitHubPullRequestNumber)
+
+	logURL, err := gc.ResolveJobLogsURL(ctx, f.FlagGitHubJob, f.FlagGitHubOwner, f.FlagGitHubRepo, f.FlagGitHubRunID)
+	if err != nil {
+		logger.WarnContext(ctx, "could not resolve direct url to job logs", "err", err)
+		logURL = fmt.Sprintf("%s/%s/%s/actions/runs/%d/attempts/%d", f.FlagGitHubServerURL, f.FlagGitHubOwner, f.FlagGitHubRepo, f.FlagGitHubRunID, f.FlagGitHubRunAttempt)
+	}
+
+	return &GitHubReporter{
+		gitHubClient: gc,
+		flags:        f,
+		logURL:       logURL,
+	}, nil
+}
+
+// validateFlags validates the required GitHub flags.
+func validateFlags(f flags.GitHubFlags) error {
```

Comment:
This feels like we're leaking an api abstraction here by re-using the flags package to share information. In reality, these are all inputs to the "GitHub Reporter" and should be "GitHubReporterInput"

Reply:
Changed to inputs